services/src/main/java/org/openlearn/web/rest/AssignmentResource.java

```diff
@@ -146,58 +165,118 @@ public ResponseEntity delete(@PathVariable final Long id) {
-		return ResponseEntity.ok(response);
+
+		AssignmentDTO assignmentDTO = assignmentService.findOne(assignmentId);
+		if (assignmentDTO == null) throw new AssignmentNotFoundException(assignmentId);
+
+		if (canUploadFilesToAssignment(assignmentDTO)) {
+			Page<FileInformationDTO> response = fileInformationService.findAllForAssignment(assignmentId, pageable);
+			return ResponseEntity.ok(response.getContent());
+		} else {
+			return new ResponseEntity(HttpStatus.FORBIDDEN);
+		}
 	}
 
-	@GetMapping(path="/{assignmentId}/upload/{keyName:.+}")
+	@GetMapping(path="/{assignmentId}/upload/{id}")
 	@Secured({AuthoritiesConstants.ADMIN, AuthoritiesConstants.ORG_ADMIN, AuthoritiesConstants.INSTRUCTOR, AuthoritiesConstants.STUDENT})
 	public ResponseEntity getUpload(@PathVariable final Long assignmentId,
-									@PathVariable final String keyName) {
-		log.debug("GET request to get course upload : {}", keyName);
-		InputStream response = storageService.getUpload(assignmentId, null, keyName);
-		if (response != null) {
-			try {
-				byte[] out = org.apache.commons.io.IOUtils.toByteArray(response);
+									@PathVariable final Long id) {
+		log.debug("GET request to get course upload : {}", id);
+		AssignmentDTO assignmentDTO = assignmentService.findOne(assignmentId);
+		FileInformationDTO fileInformationDTO = fileInformationService.findOne(id);
 
-				HttpHeaders responseHeaders = new HttpHeaders();
-				responseHeaders.add("content-disposition", "attachment; filename=" + keyName);
-				// responseHeaders.add("Content-Type", type);
+		if (assignmentDTO == null) throw new AssignmentNotFoundException(assignmentId);
+		if (fileInformationDTO == null) throw new FileInformationNotFoundException(id);
 
-				return new ResponseEntity(out, responseHeaders, HttpStatus.OK);
-			} catch (Exception e) {
-				new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
+		if (hasCreateUpdateDeleteAuthority(assignmentDTO) && fileInformationDTO.getAssignmentId() == assignmentId) {
+			String fileName = fileInformationService.getFileNameFor(id);
+			InputStream response = storageService.getUpload(id);
+			if (response != null) {
+				try {
+					byte[] out = org.apache.commons.io.IOUtils.toByteArray(response);
+
+					HttpHeaders responseHeaders = new HttpHeaders();
+					responseHeaders.add("content-disposition", "attachment; filename=" + fileName);
+					// responseHeaders.add("Content-Type", type);
+
+					return new ResponseEntity(out, responseHeaders, HttpStatus.OK);
+				} catch (Exception e) {
+					new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
+				}
+				return new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
+			} else {
+				return new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
 			}
-			return new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
 		} else {
-			return new ResponseEntity ("File Not Found", HttpStatus.NOT_FOUND);
+			return new ResponseEntity(HttpStatus.FORBIDDEN);
```

Comment:
You might want to consider returning responses/throwing exceptions earlier to avoid all of this if/else nesting.

Reply:
Agreed. Or letting the service(s) do the throwing. This is mostly legacy code which didn't have any error handling or exception, just returning `null`s so that's why there's all this if else stuff.